staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -1240,13 +1237,26 @@ func (c *cacheWatcher) add(event *watchCacheEvent, timer *time.Timer) bool {
 	}
 }
 
-func (c *cacheWatcher) nextBookmarkTime(now time.Time) (time.Time, bool) {
-	// For now we return 2s before deadline (and maybe +infinity is now already passed this time)
-	// but it gives us extensibility for the future(false when deadline is not set).
+// bookmarkHeartbeatFrequency defines how frequently bookmarks should be
+// under regular conditions.
+const bookmarkHeartbeatFrequency = time.Minute
+
+func (c *cacheWatcher) nextBookmarkTime(now time.Time) time.Time {
+	// We try to send bookmarks:
+	// (a) roughly every minutes
+	// (b) right before the watcher timeout - for now we simply set it 2s before
+	//    the deadline
+	// The former gives us periodicity if the watch breaks due to unexpected
+	// conditions, the later ensures that on timeout the watcher is as close to
+	// now as possible - this covers 99% of cases.
+	heartbeatTime := now.Add(bookmarkHeartbeatFrequency)
 	if c.deadline.IsZero() {
```

Comment:
QQ: Is it possible (watch without deadline/timeout) or is it just a safe check?

Reply:
If you're using our framework like reflector - it will always be set:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/cache/reflector.go#L387
We also try to default it it is't not provided:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/get.go#L249

But you may configure it in a way to not provide minRequestTimeout, and then it may happen (though in properly configuder custer it shouldn't happen).